src/routes/session.rs

```diff
@@ -101,3 +105,139 @@ pub fn delete(mut cookies: Cookies) -> Redirect {
     }
     Redirect::to("/")
 }
+
+#[derive(Clone)]
+pub struct ResetRequest {
+    pub mail: String,
+    pub id: String,
+    pub creation_date: Instant,
+}
+
+impl PartialEq for ResetRequest {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id
+    }
+}
+
+#[get("/password-reset")]
+pub fn password_reset_request_form(conn: DbConn, intl: I18n) -> Ructe {
+    render!(session::password_reset_request(
+        &(&*conn, &intl.catalog, None),
+        &ResetForm::default(),
+        ValidationErrors::default()
+    ))
+}
+
+#[derive(FromForm, Validate, Default)]
+pub struct ResetForm {
+    #[validate(email)]
+    pub email: String,
+}
+
+#[post("/password-reset", data = "<form>")]
+pub fn password_reset_request(
+    conn: DbConn,
+    intl: I18n,
+    mail: State<Arc<Mutex<Mailer>>>,
+    form: Form<ResetForm>,
+    requests: State<Arc<Mutex<Vec<ResetRequest>>>>
+) -> Ructe {
+    if User::find_by_email(&*conn, &form.email).is_ok() && requests.lock().unwrap().iter().find(|x| x.mail == form.email.clone()).is_none() {
+        let id = plume_common::utils::random_hex();
+        {
+            let mut requests = requests.lock().unwrap();
+            requests.push(ResetRequest {
+                mail: form.email.clone(),
+                id: id.clone(),
+                creation_date: Instant::now(),
+            });
+        }
+        // Remove outdated requests (more than 1 day old) to avoid the list to grow too much
+        let to_remove = {
```

Comment:
you should remove before checking if there is already an entry, otherwise one asking for a second reset link after, says, 12 hours, won't get it's mail. Also this mean the vec is cleaned only when a valid request is made, so if nobody else request a password reset, one wont ever get unlocked